src/pipecat/processors/audio/audio_buffer_processor.py

```diff
@@ -0,0 +1,103 @@
+import wave
+from io import BytesIO
+
+from pipecat.frames.frames import (AudioRawFrame, BotInterruptionFrame,
+                                   BotStartedSpeakingFrame,
+                                   BotStoppedSpeakingFrame, Frame,
+                                   InputAudioRawFrame, OutputAudioRawFrame,
+                                   StartInterruptionFrame,
+                                   StopInterruptionFrame,
+                                   UserStartedSpeakingFrame,
+                                   UserStoppedSpeakingFrame)
+from pipecat.processors.frame_processor import FrameDirection, FrameProcessor
+
+
+class AudioBufferProcessor(FrameProcessor):
+
+    def __init__(self):
+        """
+        Initialize the AudioBufferProcessor.
+
+        This constructor sets up the initial state for audio processing:
+        - audio_buffer: A bytearray to store incoming audio data.
+        - num_channels: The number of audio channels (initialized as None).
+        - sample_rate: The sample rate of the audio (initialized as None).
+        - assistant_audio: A boolean flag to indicate if assistant audio is being processed.
+        - user_audio: A boolean flag to indicate if user audio is being processed.
+
+        The num_channels and sample_rate are set to None initially and will be
+        populated when the first audio frame is processed.
+        """
+        super().__init__()
+        self._user_audio_buffer = bytearray()
+        self._assistant_audio_buffer = bytearray()
+        self._num_channels = None
+        self._sample_rate = None
+
+    def _buffer_has_audio(self, buffer: bytearray):
+        return (
+            buffer is not None and
+            len(buffer) > 0
+        )
+
+    def _has_audio(self):
+        return (
+            self._buffer_has_audio(self._user_audio_buffer) and
+            self._buffer_has_audio(self._assistant_audio_buffer) and
+            self._sample_rate is not None
+        )
+
+    def _reset_audio_buffer(self):
+        self._user_audio_buffer = bytearray()
+        self._assistant_audio_buffer = bytearray()
+
+    def _merge_audio_buffers(self):
+        with BytesIO() as buffer:
+            with wave.open(buffer, 'wb') as wf:
+                wf.setnchannels(2)
+                wf.setsampwidth(self._sample_rate // 8000)
```

Comment:
this should always be 2 because every sample is a signed 16-bit integer. If we change sample rate to 44100 this would break.